### PR TITLE
Add HuBERT fine-tuning recipe

### PR DESCRIPTION
**Pull request by nateanl**
The PR contains the CTC fine-tuning recipe of HuBERT Base model.
The files include:
- lightning module
- training script
- README and the result table
- evaluation scripts

### CODE REVIEW

File: examples/hubert/evaluate.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+    """
+    hypothesis = F.log_softmax(emission, dim=-1)
+    hypothesis = hypothesis.argmax(-1).unique_consecutive()
+    hypothesis = hypothesis[hypothesis != blank_idx]
+    hypothesis = "".join(id2token[int(i)] for i in hypothesis).replace("|", " ")
+    return hypothesis.split()
+
+
+def _ctc_decode(emission, decoder: CTCDecoder) -> List[str]:
+    """Run CTC decoding with KenLM language model.
```
Comment (xiaohui-zhang) [inline examples/hubert/evaluate.py:47]: with a

File: examples/hubert/evaluate.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    for idx, sample in enumerate(dataset):
+        waveform, _, transcript, _, _, _ = sample
+        transcript = transcript.strip().lower().strip().replace("\n", "")
+
+        with torch.inference_mode():
+            emission, _ = model(waveform)
+        if args.use_lm:
+            hypothesis = _ctc_decode(emission, decoder)
+        else:
+            hypothesis = _viterbi_deocde(emission, id2token)
```
Comment (xiaohui-zhang) [inline examples/hubert/evaluate.py:99]: decode

File: examples/hubert/evaluate.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    for k in state_dict:
+        if "model.wav2vec2" in k:
+            new_state_dict[k.replace("model.wav2vec2.", "")] = state_dict[k]
+        elif "aux" in k:
+            new_state_dict[k] = state_dict[k]
+    model.load_state_dict(new_state_dict)
+    return model
+
+
+def _viterbi_deocde(emission: torch.Tensor, id2token: Dict, blank_idx: int = 0) -> List[str]:
```
Comment (xiaohui-zhang) [inline examples/hubert/evaluate.py:28]: decode

File: examples/hubert/evaluate.py [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+    logger.info(f"Final WER: {total_edit_distance / total_length}")
+
+
+def _parse_args():
+    parser = argparse.ArgumentParser(
+        description=__doc__,
+        formatter_class=argparse.RawTextHelpFormatter,
+    )
+    parser.add_argument(
+        "--librispeech_path",
```
Comment (xiaohui-zhang) [inline examples/hubert/evaluate.py:115]: librispeech-path to be consistent with other args

File: examples/hubert/evaluate.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+        help="Languge model weight in decoding. (Default: 2.46)",
+    )
+    parser.add_argument(
+        "--word-score",
+        type=float,
+        default=-0.59,
+        help="Word insertion score in decoding. (Default: -0.59)",
+    )
+    parser.add_argument(
+        "--unk_score", type=float, default=float("-inf"), help="Unknown word insertion score. (Default: -inf)"
```
Comment (xiaohui-zhang) [inline examples/hubert/evaluate.py:161]: unk-score to be consistent with other args

File: examples/hubert/evaluate.py [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+    )
+    parser.add_argument("--sil_score", type=float, default=0, help="Silence insertion score. (Default: 0)")
```
Comment (xiaohui-zhang) [inline examples/hubert/evaluate.py:163]: same here

File: examples/hubert/finetune.py [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+    trainer.fit(model)
+
+
+def _parse_args():
+    parser = ArgumentParser(
+        description=__doc__,
+        formatter_class=_Formatter,
+    )
+    parser.add_argument(
+        "--root-path",
```
Comment (xiaohui-zhang) [inline examples/hubert/finetune.py:94]: dataset-path maybe?

File: examples/hubert/lightning.py [hunk 8]
```diff
@@ -43,6 +50,50 @@ def get_lr(self):
             return [base_lr * pct_remaining for base_lr in self.base_lrs]
 
 
+class TriStageLRScheduler(torch.optim.lr_scheduler._LRScheduler):
+    """Linear learning rate scheduler with warm up, hold, and decay."""
```
Comment (xiaohui-zhang) [inline examples/hubert/lightning.py:54]: warmup

File: examples/hubert/dataset/hubert_dataset.py [hunk 9]
```diff
@@ -407,3 +410,68 @@ def __call__(self, batch: List[Tuple[Tensor, Tensor, int]]) -> Tuple[Tensor, Ten
...
+
+
+class CollateFnLibriLightLimited:
+    """The collate class for LibriSpeech or LibriLightLimited dataset."""
+
+    def __call__(self, batch: Tuple[Tensor, int, str, int, int, int]) -> Tuple[Tensor, Tensor, Tensor]:
+        """
+        Args:
+            batch (Batch):
+                The list of tuples that contains
```
[the remark this answers is not in the document]
Comment (nateanl) [reply examples/hubert/dataset/hubert_dataset.py:446]: right, it's a list of tuples, will change that.

**Review by hwangjeff — APPROVED** [review]
